source/shadow-tasks/shadow_device_task.c

```diff
@@ -0,0 +1,816 @@
+    xSubscribeInfo[ 0 ].qos = MQTTQoS1;
+    /* Subscribe to shadow topic for accepted responses for submitted updates. */
+    xSubscribeInfo[ 1 ].pTopicFilter = SHADOW_TOPIC_STRING_UPDATE_ACCEPTED( democonfigCLIENT_IDENTIFIER );
+    xSubscribeInfo[ 1 ].topicFilterLength = SHADOW_TOPIC_LENGTH_UPDATE_ACCEPTED( democonfigCLIENT_IDENTIFIER_LENGTH );
+    xSubscribeInfo[ 1 ].qos = MQTTQoS1;
+    /* Subscribe to shadow topic for rejected responses for submitted updates. */
+    xSubscribeInfo[ 2 ].pTopicFilter = SHADOW_TOPIC_STRING_UPDATE_REJECTED( democonfigCLIENT_IDENTIFIER );
+    xSubscribeInfo[ 2 ].topicFilterLength = SHADOW_TOPIC_LENGTH_UPDATE_REJECTED( democonfigCLIENT_IDENTIFIER_LENGTH );
+    xSubscribeInfo[ 2 ].qos = MQTTQoS1;
+
+    /* Complete the subscribe information. The topic string must persist for
+     * duration of subscription - although in this case it is a static const so
+     * will persist for the lifetime of the application. */
+    xSubscribeArgs.pSubscribeInfo = xSubscribeInfo;
+    xSubscribeArgs.numSubscriptions = 3;
+
+    /* Loop in case the queue used to communicate with the MQTT agent is full and
+     * attempts to post to it time out.  The queue will not become full if the
+     * priority of the MQTT agent task is higher than the priority of the task
+     * calling this function. */
+    xTaskNotifyStateClear( NULL );
+    xCommandParams.blockTimeMs = shadowexampleMAX_COMMAND_SEND_BLOCK_TIME_MS;
+    xCommandParams.cmdCompleteCallback = prvSubscribeCommandCallback;
+    xCommandParams.pCmdCompleteCallbackContext = &xApplicationDefinedContext;
+    LogInfo( ( "Sending subscribe request to agent for shadow topics." ) );
+
+    do
+    {
+        /* If this fails, the agent's queue is full, so we retry until the agent
+         * has more space in the queue. */
+        xStatus = MQTTAgent_Subscribe( &xGlobalMqttAgentContext,
+                                       &( xSubscribeArgs ),
+                                       &xCommandParams );
+    } while( xStatus != MQTTSuccess );
+
+    /* Wait for acks from subscribe messages - this is optional.  If the
+     * returned value is zero then the wait timed out. */
+    ulNotificationValue = ulTaskNotifyTake( pdFALSE, pdMS_TO_TICKS( shadowexampleMS_TO_WAIT_FOR_NOTIFICATION ) );
+    configASSERT( ulNotificationValue != 0UL );
+
+    /* The callback sets the xReturnStatus member of the context. */
+    if( xApplicationDefinedContext.xReturnStatus != true )
+    {
+        LogError( ( "Failed to subscribe to shadow update topics." ) );
+        xReturnStatus = false;
+    }
+    else
+    {
+        LogInfo( ( "Successfully subscribed to shadow update topics." ) );
+    }
+
+    return xReturnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvSubscribeCommandCallback( void * pxCommandContext,
+                                         MQTTAgentReturnInfo_t * pxReturnInfo )
+{
+    bool xReturnStatus = false;
```

Comment:
readability: `xReturnStatus` is a misleading variable name for a function returning `void`.